Add v0 of versioning to AAE hashtree locks, upgrades, and object hashing

bsparrow435 opened the pull request:
This PR adds version 0 of AAE hashing to riak_kv_entropy_manager, riak_kv_index_hashtree, riak_object, and the riak_kv_vnode. Specifically, we've added logic which requires getting a hashtree lock to provide an appropriate version and then our hashtree version determines what version of riak_object:hash will be used for the incoming data.

Upgrading is completed automatically for the user. The upgrade uses a combination of riak_core_capability with a new riak_kv variable object_hash_version and logic within the riak_kv_entropy manager to intelligently upgrade hashtrees once all conditions have been met. THe upgrade flow is as follows:
1. User upgrade all nodes to 2.2.
2. Each node locally starts checking to begin upgrade in riak_kv_entropy_manager within the management tick.
3. IIF all partitions are built and have exchanged with all sibling partitions, we'll mark the riak_kv_entropy_manager state with a new pending_version.(need to exchange all siblings to catch any pre 2.1 data potentially suffering from kv679).
4. Each poke on a riak_kv_index_hashtree check the pending version of the manager and if it's set, calls the vnode to upgrade it's local tree. 
5. Vnode destroys and makes a new hashtree with the `upgrade` atom included in the start opts.
6. Once all upgraded hashtrees have registered with entropy_manager, flip our entropy_manager version to 0 to indicate this node has completed the upgrade for all local nodes.

Open areas for discussion:
1. How best to handle default case of riak_object:hash without provided version(yz hashing). Current logic flips the hash when the entropy_manager notes all local trees have upgraded.
2. hashtree:destroy only removes data within the partition dir(i.e. ./data/antie_entropy/PARTID/*). The partition ID's still need to be cleaned up out of the anti-entropy directory.

Commit:
Added logic to hash only vclock for object hashes. Added logging to catch pre 2.1 vclocks which require a rewrite to be safe

Added riak_kv boolean environment variable `hash_only_vclock` which
enables/disabled the vclock hashing logic. Additionally, when enabled,
we will no longer do the expensive compare/logging on read-repair to
catch old pre2.1 vclocks who require a rewrite.

Additional logic must be added to `yz_kv.erl` so the hashes in the
entropy_data element of the Solr documents match what is stored in KV
AAE. Upgrading the hash logic requires clearing all YZ documents and
re-indexing everything. This is because we store the obj hash in the
Solr document. This will hopefully change in the future to allow for
easier AAE changes.

Commit:
Fix two unbounded variables

Commit:
Add version to index hash tree data root

Add a static version “v0” to the data_root of the riak_kv AAE
directory. This allows us to isolate old/newer trees as well as support
downgrades which will not allow old and new trees to exchange either
locally or across MDC.

Commit:
Add version to index_hashtree state, path, and get_lock interface

New error `bad_version` added to the return from
`riak_kv_index_hashtree:do_get_lock` if the requested version and the
local version of the hash tree do not match. Requires changes in
yokozuna and riak_repl to fully function.

Commit:
Allow for version-less lock for subsystems who don't care about version

Commit:
Fix spec

Commit:
Fix specs and export arity

Commit:
Address review comments

Change boolean configuration variable to integer version starting with
0.

Change all specs from atom() to non_neg_integer() for version number.

Add extra logging for invalid non-integer hash version configurations.

Commit:
Handle cases where KV is unavailable during an exchange

Was not correctly handling case where Result list was empty of valid
object responses or only a single valid response.

Commit:
Clarity that the first element of the result is a partition Idx

Commit:
Forward merge aae hashing change v0

Add version 0 of object hashing(config `object_hash_version`) which
hashes only the vclock to represent a version of an object. Also add
versioning to index_hashtree get_lock API which now takes a Version
parameter. The absence of a version is assumed to be `undefined` and
will only be allowed to get a lock if the hash on disk is in legacy
format.

Upgraded hash trees are located within the anti_entropy data_dir under
the directory v0. Currently the only supported version is 0 but the
code allows for support of increasing integer versions.

Add verbose riak_object:equals check to get_core for all read
operations who do not trigger read repair. This logic is to detect
potentially pre-2.1 data which may be impacted by old vector clock bugs
and should be rewritten before upgrading the hash version for safety.

Commit:
WIP commit

WIP commit in case of computer failure or something like that.

Commit:
WIP push of many commits rebased. This broke some things and I'm still recovering some history

Commit:
Finish grabbing omitted final commits from previous rebase

Commit:
Fix compilation errors

Commit:
Fix pattern match, compiler issues

Also, add some testing logs and debug functions.

Commit:
Fix errors and remove some log spam by checking too frequently

Commit:
Add some debug logging

Commit:
Fix replies for vnode:command handling hash tree upgrade

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -116,16 +124,27 @@ release_lock(Pid) ->
 %%      then forwarded to the relevant index_hashtree to acquire a tree lock.
 %%      If both locks are acquired, the pid of the remote index_hashtree is
-%%      returned.
+%%      returned. This function assumes an undefined version of the hashtree
```
Instead of stating what the function does, I feel like this comment would be more helpful if it explained why it does it. Something like "this is left over for compatibility in mixed clusters with pre-2.2 nodes, from before we added versioning to the hashtree."

bsparrow435 replied:
Will change.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -189,6 +208,15 @@ expire_trees() ->
 clear_trees() ->
     gen_server:cast(?MODULE, clear_trees).
 
+get_version() ->
```
Would be nice to have type specs on these as long as we're adding new API functions. I wouldn't let it stop this from passing review since none of the other API calls in this module seem to have type specs, but just thought I'd suggest it in case you're waiting for more review comments and you have a chance to add them :)

bsparrow435 replied:
Yeah, I'll add it to all these api calls. You're right, public facing API calls should absolutely have specs.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
+    Indices = riak_core_ring:my_indices(Ring),
+    Existing = dict:from_list(VTrees),
+    MissingIdx = [Idx || Idx <- Indices,
+        not dict:is_key(Idx, Existing)],
```
If we wanted to simplify the code a little bit, we could skip the extra step of converting to a `dict` and just use `lists:keymember`.

bsparrow435 replied:
Yep absolutely, I was reusing this logic from some ring logic i added and it should be using the orddict functions.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -418,7 +453,12 @@ reload_hashtrees(Ring, State=#state{trees=Trees}) ->
     Existing = dict:from_list(Trees),
     MissingIdx = [Idx || Idx <- Indices,
                          not dict:is_key(Idx, Existing)],
-    _ = [riak_kv_vnode:request_hashtree_pid(Idx) || Idx <- MissingIdx],
```
What's the purpose of this change? Seems like the old version was simpler and would have the same effect, unless I'm missing something.

bsparrow435 replied:
This is left over from a WIP commit. I was originally doing some logic in the special empty list case before returning ok but I removed it and forgot to revert this as well.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
+    Indices = riak_core_ring:my_indices(Ring),
+    Existing = dict:from_list(VTrees),
+    MissingIdx = [Idx || Idx <- Indices,
+        not dict:is_key(Idx, Existing)],
```
Oh, or even just `orddict:is_key`, since we seem to be treating `trees_version` as an `orddict`.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -66,16 +70,20 @@
 -record(state, {mode           = automatic :: automatic | manual,
                 trees          = []        :: [{index(), pid()}],
                 tree_queue     = []        :: [{index(), pid()}],
+                trees_version  = []        :: [{index(), version()}],
```
Since we're using this as an `orddict`, we should have the type spec indicate that for clarity. i.e. `orddict:orddict(index(), version())` instead of `[{index(), version()}]`

bsparrow435 replied:
Yep, agreed, will change and do the same with trees since the code always treats trees as an orddict as well.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
+check_exchanges_and_upgrade(State) ->
+    case riak_kv_entropy_info:all_sibling_exchanges_complete() of
+        true ->
+            lager:info("Starting AAE hashtree upgrade"),
```
Should we make the level for this `notice` instead of `info`? Seems like a rare/significant enough event to warrant the extra visibility.

bsparrow435 replied:
Good feedback, will change.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
+-spec check_upgrade(state()) -> state().
+check_upgrade(State=#state{pending_version=undefined}) ->
+    State;
+check_upgrade(State=#state{pending_version=Version,trees_version=VTrees}) ->
```
Could we name this variable something like `PendingVersion` instead? It seems a little confusing to have two state fields named `version` and `pending_version` and then bind `Version` to `pending_version` and not `version`.

bsparrow435 replied:
Yep, that makes more sense, will change.

nickelization commented on src/riak_kv_entropy_manager.erl:
```diff
@@ -550,6 +598,66 @@ maybe_poke_tree(State) ->
+    case MissingIdx of
+        [] ->
+            case [Idx || {Idx, V} <- VTrees, V == Version] of
+                [] ->
```
I see the `[]` case for when none of the local hashtrees have been upgraded, and then the following case where all local trees have been upgraded, but what if only some have been upgraded? Is that a case we might run into and then end up crashing here?

bsparrow435 replied:
Good catch. This was causing some crashes in a slower test I ran last night. My normal tests allow for all trees to upgrade in the same tick for speed and was missing this error.

nickelization commented on src/riak_kv_get_core.erl:
```diff
@@ -284,6 +288,36 @@ num_pr(GetCore = #getcore{num_pok=NumPOK, idx_type=IdxType}, Idx) ->
+%% requires the user to rewrite the object in 2.1+ of Riak. Logic is disabled when using
+%% a defined version of aae hashing as it's no longer needed.
+maybe_log_old_vclock(Results) ->
+    case riak_core_capability:get({riak_kv, object_hash_version}, undefined) of
```
So right now, as I understand it, we don't start the hashtree upgrades as soon as the capability flips, because we want to make sure all partitions have exchanged first so that we get a chance to log any inconsistent data here in this function. However, it looks like we don't do any logging here once the capability flips, right? I might be missing something but it seems like we may want to be checking something other than the capability to decide whether to check for mismatched data.

bsparrow435 replied:
So I think the comment is a bit misleading and I have changed it for clarity. We only want to trigger this heavy riak_object comparison and logging when we are in the process of the hashtree upgrade. So, if the user has manually overridden the capability to set `undefined` version, we want this logic off.  If the capability has been set to 0(all nodes are upgraded to 2.2) and the riak_kv_entropy_manager:get_version call returns undefined(anything but 0) then we know the upgrade has not yet finished and we should log in specific situations where we have multiple replica objects in the Result set and they are also not equal.

nickelization commented on src/riak_kv_get_core.erl:
```diff
@@ -284,6 +288,36 @@ num_pr(GetCore = #getcore{num_pok=NumPOK, idx_type=IdxType}, Idx) ->
+%% requires the user to rewrite the object in 2.1+ of Riak. Logic is disabled when using
+%% a defined version of aae hashing as it's no longer needed.
+maybe_log_old_vclock(Results) ->
+    case riak_core_capability:get({riak_kv, object_hash_version}, undefined) of
```
Oh wait, so actually I misread the code there. The comment up above this function says "logic is disabled when using a defined version of aae hashing as it's no longer needed," but it looks like we actually disable it when we're _not_ using a defined version of aae hashing, right? Is this actually what we want to do though? Feels like something is backwards here but maybe it's just my brain hitting late Friday afternoon and starting to stall out on me ;)

nickelization commented on src/riak_kv_get_core.erl:
```diff
@@ -284,6 +288,36 @@ num_pr(GetCore = #getcore{num_pok=NumPOK, idx_type=IdxType}, Idx) ->
+            case [RObj || {_Idx, {ok, RObj}} <- Results] of
+                [] ->
+                    ok;
+                [_|[]] ->
```
Very small nitpick, but the pattern I normally see to match a singleton list is just `[_]`. I'm sure this works just as well, but feels a little bit unidiomatic.

nickelization commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -81,7 +84,8 @@
                 trees,
-                use_2i = false :: boolean()}).
+                use_2i = false :: boolean(),
+                version :: undefined | non_neg_integer()}).
```
Technically the `undefined |` is implicit in the type spec for any Erlang record field that doesn't include a default value. Maybe you already knew this and just added it for clarity, in which case I have no problem with that, but thought I'd point that out in case you'd rather cut down on the verbosity a little.

nickelization commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -198,13 +202,31 @@ get_trees({test, Pid}) ->
--spec get_lock(pid(), any(), pid()) -> ok | not_built | already_locked.
-get_lock(Tree, Type, Pid) ->
-    gen_server:call(Tree, {get_lock, Type, Pid}, infinity).
+-spec get_lock(pid(), any(), non_neg_integer(), pid()) -> ok | not_built | already_locked.
```
Should update this type spec with `bad_version` as a possible return value.

nickelization commented on src/riak_object.erl:
```diff
@@ -623,12 +623,37 @@ get_value(Object=#r_object{}) ->
+-spec vclock_hash(riak_object()) -> binary().
+vclock_hash(Obj=#r_object{}) ->
+    Hash = erlang:phash2(to_binary(v0, lists:sort(vclock(Obj)))),
+    term_to_binary(Hash).
```
As long as we're changing the hashing format, could we take this opportunity to optimize our hashing code any further?

For examples, it looks like `erlang:phash2` accepts arbitrary terms, so perhaps we don't need to wrap the result of `lists:sort(vclock(Obj))` in a call to `to_binary` - not sure if this would perform better or worse, but it seems like there's a good chance it would perform better since we'd be taking an extra encoding step out of the process.

Also, do we need to use `term_to_binary` at the end, or could we just do something more lightweight to convert it to a binary like `<<Hash:32>>`? That would almost certainly be faster, and yields smaller binaries too.

Commit:
Address review comments

Additionally, cleaned up the anode code which killed the hash tree on
upgrade. Removed one binary layer from the riak_object:hash and added
the build locks(different from concurrency lock(doh)) to the upgrade
procedure so we won’t kill trees and then have them sit empty until we
can get a build lock.

Commit:
Change some logging to debug to reduce log spam during upgrade

Commit:
Address Dialyzer Errors

Commit:
Defined orddict as a type for dialyzer

nickelization commented:
@bsparrow435 So it occurred to me this morning that we should probably have an option to disable automatically upgrading to the new tree format. It doesn't necessarily have to be documented or supported, but I think it would be good to provide an escape hatch. If nothing else, it may make testing much easier if we want to, say, write a riak_test module that starts with 2.2 and then performs a downgrade (e.g. something like `verify_riak_object_reformat`).

bsparrow435 committed:
Fix dialyzer errors from riak_repl

nickelization commented:
+1 0e6bb3e

nickelization commented:
@borshop merge
